### PR TITLE
WIP: exposes custom data through api

### PR DESCRIPTION
**Pull request by rockitbaby**
arbitrary data unter the key `custom` in meta.yaml or meta.json is accessible via the api

This will enable creators of custom frontends to use custom data fields / values / configuration defined in metadata. Data needs to be evaluated and validated in the frontend.

Please review, @mariuswilms.

Plannend next steps:

- [x] update example data to demon custom in metadata
- [x] extend frontend to show custom[synonyms] when custom is not null and synonyms is of kind array
- [x] fix yaml unmarshalling (see comment below)
- [ ] update documentation

### CODE REVIEW

**Comment by rockitbaby** [comment]
just realized that yaml to json needs more work.

```yaml
custom:
  synonyms:
    - Listing
    - List
```

produces an map[interface{}]interface{} which is not handled by json Marshal

https://stackoverflow.com/questions/40737122/convert-yaml-to-json-without-struct/40737676#40737676

may i need to write a converter function?

**Comment by rockitbaby** [comment]
example with json works well

**Comment by mariuswilms** [comment]
In theory (I've just read what you've linked to) the converter function from https://github.com/icza/dyno/blob/master/dyno.go#L563-L601 applied to YAML after unmarshalling should do it. Feel free to include the whole library, its tiny enough.

**Comment by rockitbaby** [comment]
thank you @mariuswilms - added dyno and now use ConvertMapI2MapS when yaml file is loaded

**Comment by francawinter** [comment]
@rockitbaby I will try to develop a generic frontend solution in addition to your backend work. I will tell you when it's done and then we'll see how we merge it together. – Franca

File: apiv1.go [hunk 1]
```diff
@@ -263,7 +264,7 @@ func (api APIv1) NewNode(n *Node) (*APIv1Node, error) {
 		Related:     related,
 		Prev:        prev,
 		Next:        next,
-
+		Custom:		 n.Custom(),
```
Comment (mariuswilms) [inline apiv1.go]: Lets keep a new line to separate the deprecated fields from the rest. It also looks like you'll need to run this through `gofmt`, but maybe this is GitHub rendering the diff oddly (the field value seems to be indented to far).

File: example/02_Components/Displaying Content/List/meta.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    "Release/0.6"
+  ],
+  "version": 1,
+  "custom": {
+    "synonyms": [
+      "Listing",
+      "Index"
+    ]
+  }
+}
```
Comment (mariuswilms) [inline example/02_Components/Displaying Content/List/meta.json]: I think this is a leftover from an inbetween stage, when there was not yet YAML support on the feature. Can you restore the YAML (below) and add the custom keys there, then remove the JSON representation?

File: node.go [hunk 3]
```diff
@@ -218,6 +218,11 @@ func (n *Node) Description() string {
 	return n.meta.Description
 }
 
+// Returns the full description of the node.
```
Comment (mariuswilms) [inline node.go]: The documentation seems to be wrong. Although we try to document as many exported functions as possible, we do not try to cover them all. I think you safely remove the comment.

File: node_meta.go [hunk 4]
```diff
@@ -59,8 +61,12 @@ func (m *NodeMeta) Load() error {
 	case ".json":
 		return json.Unmarshal(contents, &m)
 	case ".yaml", ".yml":
-		return yaml.Unmarshal(contents, &m)
+		if err := yaml.Unmarshal(contents, &m); err != nil {
+			return err
+		}
+		m.Custom = dyno.ConvertMapI2MapS(m.Custom)
+		return nil
```
Comment (mariuswilms) [inline node_meta.go]: :+1:

File: node_meta.go [hunk 5]
```diff
@@ -59,8 +61,12 @@ func (m *NodeMeta) Load() error {
...
 	default:
 		return fmt.Errorf("Unsupported format: %s", prettyPath(m.path))
 	}
-}
+}
```
Comment (mariuswilms) [inline node_meta.go]: Have `gofmt` clean this up.

File: node.go (hunk 3 above)
Comment (rockitbaby) [reply node.go]: oops, fixed c5d900f

File: apiv1.go (hunk 1 above)
Comment (rockitbaby) [reply apiv1.go]: done: 2f1acd1 - also: used gofmt for the first time 🎉

File: example/02_Components/Displaying Content/List/meta.json (hunk 2 above)
Comment (rockitbaby) [reply example/02_Components/Displaying Content/List/meta.json]: done, was not sure, but decided to keep the custom data example

**Comment by mariuswilms** [comment]
Thanks for working on this, I've squashed and merged your commits as ab98c83abffe0ec331e5938c5696b15a6e8056db

**Pull request closed without merging**